### ISSUE TITLE
Encyclopedia for the AI

### ISSUE THREAD

**Issue by MikkoMMM** [cloveranon/Clover-Edition#95]
Improving the AI's contextual understanding is something often requested for AI Dungeon. One idea I saw floating around is that there could be a storage of facts from which the game could draw from when they are relevant, and input those facts to the AI when it's creating a response to the player. They talked about language parsing for new facts, but I think I'm being more realistic with a system of manual inputting of facts and the names by which they are known. So here's my idea for an "encyclopedia for the AI":

Entries:
- Each entry would have a primary call name by which it's listed, and the finding of which in the user input always triggers adding the encyclopedia entry into the input going to the AI.
- Each entry also has zero or more synonyms. These are for grammatical cases, possibly personal pronouns, and other words that trigger the entry.
- If two or more different entries are called with the same synonym, the game first tries to determine which one(s) have been called with their primary call name(s), and then leave out the other entries. If, however, no primary call name has been used in the prompt, the game asks as to which entry/entries the player is referring to. Presumably if the player is using the feature in this manner, they won't mind.
- Then of course, there's the description itself for the entry. This is what is added to the input going to the AI when the entry becomes relevant according to the previous rules.
- There should also be a setting to toggle between "always on", "contextual" (what I described), and "always off" modes.

Categories:
- I think that categories for entries is also a good idea since for example if the player tends to add personal pronouns for a lot of peoples' synonyms, they would be asked very often for clarification if there are no higher level toggles. Personally I think one level of categorization might be good enough. However, I'm not necessarily against future-proofing with a hierarchy, either.
- As with entries, the categories are named by the player. However, they only have toggles between "on" and "off". The "on" mode respects each individual entry's toggle, while the "off" mode makes the whole category be not considered for addition.
- Entries can inherit synonyms from their parent category. This way, if one really wants to play that way, one can for example have the categories for males and females in the story so they don't need to list all the relevant personal pronouns for every single entry

But would all this just be more trouble than it's worth to a player? The AI doesn't operate within normal boundaries of logic and may just ignore a lot of the "facts" sometimes, or selectively pick words from there and use them in something completely different. However, one possible advantage of this is that we may be able to reduce the overhead to AI's response time that a long context can (I think at least) cause, since now only parts of what is stored permanently are applied every time the AI generates a response.

Another question worth asking is: do we have the means to implement something like that? It sounds like a fairly sizable feature at least to me. I think that the Prompt Toolkit may also have some kind of checkbox functionality, so all those toggles would be easier with it. I haven't given much thought into the user interface for this with the classic input system.

**Comment by Akababa** [comment]
I think this is more or less the idea behind the self-attention mechanism that GPT2 is based on, where every input token is an "encyclopedia entry" that the attention mechanism decides how much to weight to give. Then with multiple layers of transformers, the NN builds up a more abstract input representation which it can use to store whatever information it wants.

Of course, that's not to say we can't do better with some preprocessing/feature engineering to help the AI, but it would require some form of clever retraining. For example, we could allow it to store info between curly braces (such as {John is your best friend}) that isn't counted in the output, to give it more memory than it already has in the input sequence, but there are already so many layers and parameters in GPT2 that I'm sure it could store this information in one of the intermediate representations of the model if it wanted to.

**Comment by MikkoMMM** [comment]
I heard some discussion about the GPT2 works. It doesn't store anything; it just takes an input. When I look at the code, I see following two types of textual input to it:
```
            self.get_story() + action,
            self.context + ' '.join(self.memory),
```
I think that it also limits the memories it has. There's at least a max_history_tokens variable in the generator code. Therefore, in longer games, stuff can just be forgotten by the AI, even if were to be trained to "perfection" so that it for all intents and purposes seemingly understood what was going on. Contextual memory would help it retain information for longer games when it becomes relevant once more.

You probably had something else in mind when you said that but the curly brace idea would sound quite user-friendly for quickly adding encyclopedia entries. Though, it wouldn't add the corresponding synonyms for the entry.

**Comment by DonaldTsang** [comment]
https://github.com/AIDungeon/AIDungeon/issues/181

**Comment by DonaldTsang** [comment]
You might want to read up on Knowledge-Graph supported models in https://github.com/thunlp/PLMpapers for some ideas as to how GPT-2 can be enhanced.

**Comment by cloveranon** [comment]
@MikkoMMM did you invent world info?
People liked that. I prefer to just manually input all relevant facts into the AI's memory and manually mange them. The real problem is the AI's memory is very limited.
Possibly you could *briefly* train/finetune the model on your list of generated facts. Like a tiny learning rate applied on like 2 training steps. It could perhaps be done so it memorizes those facts and "knows" them in some sense, while still retaining most of the information it learned before.

**Comment by MikkoMMM** [comment]
> @MikkoMMM did you invent world info?
No idea, but it's not rare for people to have similar ideas independently.

Hmm, that idea about finetuning the AI on something is indeed different. I kind of like the idea since there's not really much of a limit to how much data you can input then. The process has to be made streamlined, though, so that the user doesn't require any knowledge of the process and it must be done in memory, I'd say. Do you think the results would be different whether it's all finetuned at once (when loading a save) or when it's fed facts one at a time?

**Comment by cloveranon** [comment]
I think we should just add support for finetuning (as an extra script or something) and let adventurous users experiment. If anything works we can make it a fully supported feature. I have a basic test script running, I'll work on cleaning it up and publishing it this week.
